core/src/main/java/software/amazon/awssdk/core/async/FileAsyncRequestBody.java

```diff
@@ -186,13 +210,15 @@ public void completed(Integer result, ByteBuffer attachment) {
                         position += attachment.remaining();
                         subscriber.onNext(attachment);
                         // If we have more permits, queue up another read.
-                        if (outstandingRequests.decrementAndGet() > 0) {
+                        if (outstandingDemand.decrementAndGet() > 0) {
                             readData();
                             return;
                         }
                     } else {
                         // Reached the end of the file, notify the subscriber and cleanup
-                        subscriber.onComplete();
+                        synchronized (subscriber) {
```

Comment:
Why does this need to be synchronized?

Reply:
Interactions with the subscriber need external synchrnozation. Without it, it's possible for a race condition between `request(-1)` triggering `onError` and `this call to `onComplete` from the handler thread

Reply:
Do we have any logic that would prevent both onError and onComplete from being called on the same subscriber? Even with synchronization that can still happen right?

Reply:
Yeah I think that can with this implementation but I think that's a separate issue.

The purpose of this synchronization is 1.3:

> onSubscribe, onNext, onError and onComplete signaled to a Subscriber MUST be signaled in a thread-safe manner—and if performed by multiple threads—use external synchronization.

I think the rule you're referring to is 1.7:

> Once a terminal state has been signaled (onError, onComplete) it is REQUIRED that no further signals occur.

We'd need additional state tracking to comply with 1.7

Reply:
I see. So does the TCK not test the 1.7 requirement?

Reply:
I guess not. I'll go ahead and add some state tracking here

Reply:
Found these: https://github.com/reactive-streams/reactive-streams-jvm/blob/43fd6d18d45484575facf42b82f5e6bffe35a486/tck/src/main/java/org/reactivestreams/tck/PublisherVerification.java#L428-L447

But it looks like one is not implemented at the moment, and there's no test that checks signaling `onError` after `onComplete` or vice versa